lib/execution_segment.go

```diff
@@ -293,3 +300,97 @@ func (es *ExecutionSegment) CopyScaleRat(value *big.Rat) *big.Rat {
 	}
 	return new(big.Rat).Mul(value, es.length)
 }
+
+// ExecutionSegmentSequence represents an ordered chain of execution segments,
+// where the end of one segment is the beginning of the next. It can serialized
+// as a comma-separated string of rational numbers "r1,r2,r3,...,rn", which
+// represents the sequence (r1, r2], (r2, r3], (r3, r4], ..., (r{n-1}, rn].
+// The empty value should be treated as if there is a single (0, 1] segment.
+type ExecutionSegmentSequence []*ExecutionSegment
+
+// NewExecutionSegmentSequence validates the that the supplied execution
+// segments are non-overlapping and without gaps. It will return a new execution
+// segment sequence if that is true, and an error if it's not.
+func NewExecutionSegmentSequence(segments ...*ExecutionSegment) (ExecutionSegmentSequence, error) {
+	if len(segments) > 2 {
+		to := segments[0].to
+		for i, segment := range segments[1:] {
+			if segment.from.Cmp(to) != 0 {
+				return nil, fmt.Errorf(
+					"the start value %s of segment #%d should be equal to the end value of the previous one, but it is %s",
+					segment.from, i+1, to,
+				)
+			}
+			to = segment.to
+		}
+	}
+	return ExecutionSegmentSequence(segments), nil
+}
+
+// NewExecutionSegmentSequenceFromString parses strings of the format
+// "r1,r2,r3,...,rn", which represents the sequences like (r1, r2], (r2, r3],
+// (r3, r4], ..., (r{n-1}, rn].
+func NewExecutionSegmentSequenceFromString(strSeq string) (ExecutionSegmentSequence, error) {
+	var segments []*ExecutionSegment
+	if len(strSeq) != 0 {
+		points := strings.Split(strSeq, ",")
+		if len(points) < 2 {
+			return nil, fmt.Errorf("at least 2 points are needed for an execution segment sequence, %d given", len(points))
+		}
+		start := points[0]
+
+		segments = make([]*ExecutionSegment, 0, len(points)-1)
+		for _, point := range points[1:] {
+			segment, errl := NewExecutionSegmentFromString(start + ":" + point)
```

Comment:
if we had a "construtor" taking two `big.Rat`s I would've preferred if we directly called stringToRat on `point` instead of here making a string that then gets decompositions in the new "constructor", but given that we don't I just give it as possible optimization

Reply:
We're not dealing with rational numbers exclusively, this can be `0.50-0.75`

Reply:
Ah, sorry, my bad, `stringToRat` will deal with that just fine. So yeah, :+1: for this approach, it would be better.

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1318/commits/686f32b9204b02f4214be812ad2ec1966b4463fc -